RFD.md

```diff
@@ -0,0 +1,217 @@
+
+* PID namespace - isolates user commands from the host OS process namespace to avoid leaking process information from the host into the container.
+
+* Network namespace - isolates container networking stack from that of the host OS.
+
+* Mount namespace - allows us to fully isolate command's filesystem from that of the host OS (and protect host OS `/proc` from being accessed by the user command to gain access to any privileged information on the host).
+
+##### Scope limits
+
+For this exercise, we're going to accept the following constraints:
+
+* Networking namespace will be used to separate the user command from the host OS network, but we're not going to build any bridges, etc to connect container network to the host.
+
+* We're going to create a new clean filesystem by extracting an Alpine Root FS image per command, which will use 5Mb of disk space per container. In a production environment we would probably rely on some kind of layered filesystem like overlayfs to save on disk resources.
+
+#### Availability and resource limits
+
+In a multi-tenant environment like the system in question, we need to ensure proper resource limits to reduce the potential impact of noisy or abusive clients on the underlying host OS.
+
+We're going to use linux Cgroups to apply the following limits on each command:
+
+* Memory usage – a static per-command limit will be applied on each container.
+
+* CPU usage - each command will be limited to a small constant amount of CPU resources, preventing it from negative affecting the host OS and other commands by consuming too much CPU.
+
+* Disk IO - each command will be limited by the amount of disk IO it could perform to limit the effects from a single command's high IO load on the host OS and other commands.
+
+##### Scope limits
+
+All limits will be statically hardcoded within the server, while in a production system we'd probably have some kind of configuration file for default limits and would expand the API to allow users to specify their own limits for each command (like Docker does).
+
+In a production environment, we may want to introduce per-user resource limits, API call rate limits, etc to further protect the system from abuse.
+
+### API
+
+The following API methods will be available on the server:
+
+* `Status` - returns some basic status information
+* `StartCommand` - starts a command on the server and returns a unique `command_id` value used to manage the command in subsequent API calls.
+* `StopCommand` - stops a given command (identified by a `command_id`).
+* `CommandStatus` - returns current status information for a given command (identified by a `command_id`).
+* `CommandOutput` - a streaming API for receiving console output from a given command (identified by a `command_id`). If requested, the stream will continue until the command has finished (tail mode).
+
+The detailed GRPC definition for the proposed API can be found within the `remote_exec.proto` file.
```

Comment:
Can you please include the `remote_exec.proto` file? It seems to be missing.

Reply:
Doh! Added now.